## Python Programming Project: Profitable App Data Profiling
---

### Description
- This project is designed to demonstrate looping and data profiling skills gained from Dataquest.io training.  
- The perspective taken is that of a company that develops Android and iOS apps for the Apple App Store.
- GOAL: to understand the market for free apps, which is the business model of our ficticious company.

#### Assumptions:
- Our company is targeting the US (only cares about English language apps)

### Data
This data set is from apps that were downloaded, how much they cost, and also ratings of users.  Other metrics are provided to understand how successful apps are across multiple systems.
- The [source](https://www.kaggle.com/ramamet4/app-store-apple-data-set-10k-apps/data) of the Apple app data is from a Kaggle competition.
- The [source](https://www.kaggle.com/lava18/google-play-store-apps) of the Android app data is from another Kaggle competition.

### Key 

| iOS Name | Android Name |Description |
| ----------- | ----------- |---------------------- |
| "id" | NA | App ID |
| "track_name" |"App" |  App Name |
| "prime_genre" | "Category" | Category the app belongs to |
| "size_bytes" |"Size" | Size (in Bytes) |
| "currency" | NA | Currency of price |
| "price" |"Price" | Purchase price of the app |
| "rating_count_tot" |"Reviews" | Total count of ratings |
| "user_rating" |"Rating" | Average of user ratings |
| "cont_rating" |"Content Rating" | Age group the app is targeted at |
| "rating_count_ver" | NA | Count of current version ratings |
| "user_rating_ver" | NA | Average user ratings for current version |
| "sup_devices.num" | NA | Unknown |
| "ipadSc_urls.num" | NA | Unknown |
| "lang.num" | NA | Unknown |
| "vpp_lic" | NA | Unknown |
| NA | "Genres" | An app can belong to multiple genres (apart from its main category)|
| NA | "Installs" | Number of user downloads/installs for the app (as when scraped) |
| NA | "Last Updated" | Date when the app was last updated on Play Store (as when scraped) |
| NA | "Current Ver" | Current version of the app available on Play Store (as when scraped) |
| NA | "Android Ver" | Min required Android version (as when scraped) |

----

#### Reading in the data
We have already saved the csv files from the locations above. We will read each into separate files, since they contain different data elements

In [11]:
from csv import reader
apple_file = open('AppleStore.csv', encoding='utf8')
android_file = open('googleplaystore.csv', encoding='utf8')
read_apple_file = reader(apple_file)
read_android_file = reader(android_file)

ios = list(read_apple_file)
ios_header = ios[0]
ios = ios[1:]

android = list(read_android_file)
android_header = android[0]
android = android[1:]

#### Explore the data
Let's take some time looking at the headers and some row examples of the iOS and Android apps. We will have to do this separately because we have not merged or unioned them together yet.

In [12]:
def explore_data(dataset, start, end, rows_and_columns=False):
    dataset_slice = dataset[start:end]    
    for row in dataset_slice:
        print(row)
        print('\n') # adds a new (empty) line after each row

    if rows_and_columns:
        print('Number of rows:', len(dataset))
        print('Number of columns:', len(dataset[0]))

In [17]:
print(ios_header)

['', 'id', 'track_name', 'size_bytes', 'currency', 'price', 'rating_count_tot', 'rating_count_ver', 'user_rating', 'user_rating_ver', 'ver', 'cont_rating', 'prime_genre', 'sup_devices.num', 'ipadSc_urls.num', 'lang.num', 'vpp_lic']


In [13]:
explore_data(ios,0,2, True)

['1', '281656475', 'PAC-MAN Premium', '100788224', 'USD', '3.99', '21292', '26', '4', '4.5', '6.3.5', '4+', 'Games', '38', '5', '10', '1']


['2', '281796108', 'Evernote - stay organized', '158578688', 'USD', '0', '161065', '26', '4', '3.5', '8.2.2', '4+', 'Productivity', '37', '5', '23', '1']


Number of rows: 7197
Number of columns: 17


In [16]:
print(android_header)

['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']


In [14]:
explore_data(android,0,2, True)

['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['Coloring book moana', 'ART_AND_DESIGN', '3.9', '967', '14M', '500,000+', 'Free', '0', 'Everyone', 'Art & Design;Pretend Play', 'January 15, 2018', '2.0.0', '4.0.3 and up']


Number of rows: 10841
Number of columns: 13


#### Deleting wrong data
From the discussion boards, we find out that there is a row that missing the category.

In [20]:
print(android[10472])

['osmino Wi-Fi: free WiFi', 'TOOLS', '4.2', '134203', '4.1M', '10,000,000+', 'Free', '0', 'Everyone', 'Tools', 'August 7, 2018', '6.06.14', '4.4 and up']


In [19]:
# del android[10472] # only do this once

#### Handling duplicate data
There are some duplicate rows - apps that are repeated in the data set. An example is shown below and we will need to filter out the ones with less than the max number of reviews.

---

In [21]:
for app in android:
    name = app[0]
    if name == 'Instagram':
        print(app)

['Instagram', 'SOCIAL', '4.5', '66577313', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']
['Instagram', 'SOCIAL', '4.5', '66577446', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']
['Instagram', 'SOCIAL', '4.5', '66577313', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']
['Instagram', 'SOCIAL', '4.5', '66509917', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']


In [23]:
duplicate_apps = []
unique_apps = []
for app in android:
    name = app[0]
    if name in unique_apps:
        duplicate_apps.append(name)
    else:
        unique_apps.append(name)
print('Number of duplicate apps: ',len(duplicate_apps))
print('\n')
print('Examples of duplicate apps: ', duplicate_apps[:15])

Number of duplicate apps:  1181


Examples of duplicate apps:  ['Quick PDF Scanner + OCR FREE', 'Box', 'Google My Business', 'ZOOM Cloud Meetings', 'join.me - Simple Meetings', 'Box', 'Zenefits', 'Google Ads', 'Google My Business', 'Slack', 'FreshBooks Classic', 'Insightly CRM', 'QuickBooks Accounting: Invoicing & Expenses', 'HipChat - Chat Built for Teams', 'Xero Accounting Software']


---
We won't be removing duplicates randomly. These aren't true duplicates - they are simply updates of existing apps that have increasing amounts of reviews. We will at least be removing based on number of reviews, choosing the maximum review count record to keep.

---

In [28]:
reviews_max = {}
for row in android:
    app_name = row[0]
    n_reviews = float(row[3])
    if app_name in reviews_max and reviews_max[app_name] < n_reviews:
        reviews_max[app_name] = n_reviews
    elif app_name not in reviews_max:
        reviews_max[app_name] = n_reviews

Let's check to see if we have removed the correct amount of duplicates.

In [32]:
eval_len = len(android)-1181 == len(reviews_max)
print('Have the records been shortened by 1181 duplicates? ', eval_len)

Have the records been shortened by 1181 duplicates?  True


Let's use the dictionary we created as we loop through the android data set.  By the end we will have a new data set that has the duplicate rows removed.  As we loop, we will:

1. Add the apps that have the highest reviews (refering to the dictionary)
2. Deep track if we had already added them.

This logic is captured in the if statement

In [47]:
android_clean = []
already_added = []
for row in android:
    name = row[0]
    n_reviews = float(row[3])
    if n_reviews == reviews_max[name] and name not in already_added:
        android_clean.append(row)
        already_added.append(name)

In [45]:
print('The expected length of the new list is: ',len(reviews_max))
print('The actual length of the new list is: ', len(android_clean))

The expected length of the new list is:  9659
The actual length of the new list is:  9659


The length matches our expectation of 9659, we are good to proceed to the next step.

---

#### Filter non-English language apps
Since our company is targeting the US market, we will need to confirm that no non-English apps exist in the data set.  We will do this by removing each app with a name containing symbol(s) not commonly used in English.

Fortunately, we can use the ASCII system range of 0 to 127 which corresponds to those of English characters.  In other words, if an app name contains a character greater than 127, then it is probably not an English name.

However, sometimes there are emoji's or trademarks that fall outside the 127 range.  Since we don't want to filter these out, we will add some logic that allows up to 3 non-English characters.  A list of non-English characters will keep track of this count as we loop through all the characters in the string.

In [59]:
def english_checker(string):
    is_english = True
    non_english_characters = []
    for character in string:
        ascii_num = ord(character)
        if ascii_num > 127:
            non_english_characters.append(ascii_num)
    if len(non_english_characters) > 3:
        is_english = False
    return is_english

Let's see how this function performs:

In [60]:
english_checker('Instagram')

True

In [61]:
english_checker('爱奇艺PPS -《欢乐颂2》电视剧热播')

False

In [62]:
english_checker('Docs To Go™ Free Office Suite')

True

In [63]:
english_checker('Instachat 😜')

True

Now let's use this function to loop through our two data sets to filter out non-English apps.  If an app name is English, we will add it to a new list.

In [64]:
android_english = []
for row in android_clean:
    if english_checker(row[0]):
        android_english.append(row)

In [65]:
explore_data(android_english,0,2, True)

['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


Number of rows: 9614
Number of columns: 13


In [66]:
ios_english = []
for row in ios:
    if english_checker(row[2]):
        ios_english.append(row)

In [67]:
explore_data(ios_english,0,2, True)

['1', '281656475', 'PAC-MAN Premium', '100788224', 'USD', '3.99', '21292', '26', '4', '4.5', '6.3.5', '4+', 'Games', '38', '5', '10', '1']


['2', '281796108', 'Evernote - stay organized', '158578688', 'USD', '0', '161065', '26', '4', '3.5', '8.2.2', '4+', 'Productivity', '37', '5', '23', '1']


Number of rows: 6183
Number of columns: 17


In [68]:
### number of columns for iOS doesn't match solutions. May have kept index column

### Filter to only free apps
Our company only builds apps that are free to download, thus we want to limit our market evaluation to free apps. Below we loop through both lists and retain only the ones with zero for price.

In [84]:
ios_eng_free = []
for row in ios_english:
    price = row[5]
    if price == '0':
        ios_eng_free.append(row)
explore_data(ios_eng_free,0,1,True)

['2', '281796108', 'Evernote - stay organized', '158578688', 'USD', '0', '161065', '26', '4', '3.5', '8.2.2', '4+', 'Productivity', '37', '5', '23', '1']


Number of rows: 3222
Number of columns: 17


In [72]:
android_eng_free = []
for row in android_english:
    price = row[7]
    if price == '0':
        android_eng_free.append(row)
explore_data(android_eng_free,0,1,True)

['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


Number of rows: 8864
Number of columns: 13


### Choose a genre to target
The aim of our company is to earn revenue from advertising.  Thus the more users, the better. Our strategy is as follows:
1. Make an minimal viable product for Android, add it to Google Play
2. If the app does well, develop more features
3. If the app is profitable after 6 months, then we build an iOS version.

Hence, we want to understand which genre's exist in both stores and how popular they might be.
Let's build a function to allow a dataset and column index as inputs and spit out a percentage frequency table.

In [100]:
def freq_table(dataset, index):
    
# Create a dictionary with counts for each uniqe column value
    column_counting = {}
    for row in dataset:
        column = row[index]
        if column in column_counting:
            column_counting[column] += 1
        else:
            column_counting[column] = 1
        
# Convert the counts to percentages
    total = len(dataset)
    percentages = {}
    for key in column_counting:
        percentage = 100 * column_counting[key]/total
        percentage = round(percentage,2)
        percentages[key] = percentage
        
    return percentages

In [101]:
ios_genre_percentages = freq_table(ios_eng_free, 12)
print(ios_genre_percentages)

{'Productivity': 1.74, 'Weather': 0.87, 'Shopping': 2.61, 'Reference': 0.56, 'Finance': 1.12, 'Music': 2.05, 'Utilities': 2.51, 'Travel': 1.24, 'Social Networking': 3.29, 'Sports': 2.14, 'Health & Fitness': 2.02, 'Games': 58.16, 'Food & Drink': 0.81, 'News': 1.33, 'Book': 0.43, 'Photo & Video': 4.97, 'Entertainment': 7.88, 'Business': 0.53, 'Lifestyle': 1.58, 'Education': 3.66, 'Navigation': 0.19, 'Medical': 0.19, 'Catalogs': 0.12}


In [102]:
android_genre_percentages = freq_table(android_eng_free, 1)
print(android_genre_percentages)

{'ART_AND_DESIGN': 0.64, 'AUTO_AND_VEHICLES': 0.93, 'BEAUTY': 0.6, 'BOOKS_AND_REFERENCE': 2.14, 'BUSINESS': 4.59, 'COMICS': 0.62, 'COMMUNICATION': 3.24, 'DATING': 1.86, 'EDUCATION': 1.16, 'ENTERTAINMENT': 0.96, 'EVENTS': 0.71, 'FINANCE': 3.7, 'FOOD_AND_DRINK': 1.24, 'HEALTH_AND_FITNESS': 3.08, 'HOUSE_AND_HOME': 0.82, 'LIBRARIES_AND_DEMO': 0.94, 'LIFESTYLE': 3.9, 'GAME': 9.72, 'FAMILY': 18.91, 'MEDICAL': 3.53, 'SOCIAL': 2.66, 'SHOPPING': 2.25, 'PHOTOGRAPHY': 2.94, 'SPORTS': 3.4, 'TRAVEL_AND_LOCAL': 2.34, 'TOOLS': 8.46, 'PERSONALIZATION': 3.32, 'PRODUCTIVITY': 3.89, 'PARENTING': 0.65, 'WEATHER': 0.8, 'VIDEO_PLAYERS': 1.79, 'NEWS_AND_MAGAZINES': 2.8, 'MAPS_AND_NAVIGATION': 1.4}


In [103]:
def display_table(dataset, index):
    table = freq_table(dataset, index)
    table_display = []
    for key in table:
        key_val_as_tuple = (table[key], key)
        table_display.append(key_val_as_tuple)

    table_sorted = sorted(table_display, reverse = True)
    for entry in table_sorted:
        print(entry[1], ':', entry[0])

In [104]:
display_table(ios_eng_free, 12)

Games : 58.16
Entertainment : 7.88
Photo & Video : 4.97
Education : 3.66
Social Networking : 3.29
Shopping : 2.61
Utilities : 2.51
Sports : 2.14
Music : 2.05
Health & Fitness : 2.02
Productivity : 1.74
Lifestyle : 1.58
News : 1.33
Travel : 1.24
Finance : 1.12
Weather : 0.87
Food & Drink : 0.81
Reference : 0.56
Business : 0.53
Book : 0.43
Navigation : 0.19
Medical : 0.19
Catalogs : 0.12


For free english apps, Games accounts for the majority fo the selection at 58%. Entertainment is near 8% and Photo & Video follows at 5%. It will be interesting to see if the number of apps yields the most users or if some of the other more "boring" genre's like News, Finance, Reference, or Business effectively have many more user's per app.

In [105]:
display_table(android_eng_free, 1)

FAMILY : 18.91
GAME : 9.72
TOOLS : 8.46
BUSINESS : 4.59
LIFESTYLE : 3.9
PRODUCTIVITY : 3.89
FINANCE : 3.7
MEDICAL : 3.53
SPORTS : 3.4
PERSONALIZATION : 3.32
COMMUNICATION : 3.24
HEALTH_AND_FITNESS : 3.08
PHOTOGRAPHY : 2.94
NEWS_AND_MAGAZINES : 2.8
SOCIAL : 2.66
TRAVEL_AND_LOCAL : 2.34
SHOPPING : 2.25
BOOKS_AND_REFERENCE : 2.14
DATING : 1.86
VIDEO_PLAYERS : 1.79
MAPS_AND_NAVIGATION : 1.4
FOOD_AND_DRINK : 1.24
EDUCATION : 1.16
ENTERTAINMENT : 0.96
LIBRARIES_AND_DEMO : 0.94
AUTO_AND_VEHICLES : 0.93
HOUSE_AND_HOME : 0.82
WEATHER : 0.8
EVENTS : 0.71
PARENTING : 0.65
ART_AND_DESIGN : 0.64
COMICS : 0.62
BEAUTY : 0.6


For Android, there is a much more even distribution. Nothing dominates like Games did for iOS. The highest category is Family at 19% followed by Games and Tools near 8-9%. However, keep in mind that Family describes apps for kids (which maybe classified as Games on iOS).

In [106]:
display_table(android_eng_free, 9)

Tools : 8.45
Entertainment : 6.07
Education : 5.35
Business : 4.59
Productivity : 3.89
Lifestyle : 3.89
Finance : 3.7
Medical : 3.53
Sports : 3.46
Personalization : 3.32
Communication : 3.24
Action : 3.1
Health & Fitness : 3.08
Photography : 2.94
News & Magazines : 2.8
Social : 2.66
Travel & Local : 2.32
Shopping : 2.25
Books & Reference : 2.14
Simulation : 2.04
Dating : 1.86
Arcade : 1.85
Video Players & Editors : 1.77
Casual : 1.76
Maps & Navigation : 1.4
Food & Drink : 1.24
Puzzle : 1.13
Racing : 0.99
Role Playing : 0.94
Libraries & Demo : 0.94
Auto & Vehicles : 0.93
Strategy : 0.91
House & Home : 0.82
Weather : 0.8
Events : 0.71
Adventure : 0.68
Comics : 0.61
Beauty : 0.6
Art & Design : 0.6
Parenting : 0.5
Card : 0.45
Casino : 0.43
Trivia : 0.42
Educational;Education : 0.39
Board : 0.38
Educational : 0.37
Education;Education : 0.34
Word : 0.26
Casual;Pretend Play : 0.24
Music : 0.2
Racing;Action & Adventure : 0.17
Puzzle;Brain Games : 0.17
Entertainment;Music & Video : 0.17
Casual;

At first glance, the columns of Category and Genres look very similar, but Genres does contain some more detailed categories not listed in the Categories column.  Despite this, we will stay at the higher level Categories column for analysis.

### Which apps have the most users
To determine which genre (or category) has the most users is to calculate the average number of downloads (or reviews when not available) per app genre. 